ambry-commons/src/main/java/com.github.ambry.commons/InputStreamReadableStreamChannel.java

```diff
@@ -0,0 +1,225 @@
+      FutureResult<Long> futureResult = new FutureResult<Long>();
+      futureResult.done(0L, closedChannelException);
+      future = futureResult;
+      if (callback != null) {
+        callback.onCompletion(0L, closedChannelException);
+      }
+    } else if (!readIntoCalled.compareAndSet(false, true)) {
+      throw new IllegalStateException("ReadableStreamChannel cannot be read more than once");
+    } else {
+      callbackWrapper = new ReadIntoCallbackWrapper(callback);
+      future = callbackWrapper.futureResult;
+      if (size == 0) {
+        callbackWrapper.invokeCallback(null);
+      } else {
+        writeToChannel(asyncWritableChannel);
+      }
+    }
+    return future;
+  }
+
+  @Override
+  public boolean isOpen() {
+    return channelOpen.get();
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * Also closes the underlying {@link InputStream}.
+   * @throws IOException
+   */
+  @Override
+  public void close()
+      throws IOException {
+    if (channelOpen.compareAndSet(true, false)) {
+      inputStream.close();
+    }
+  }
+
+  /**
+   * Reads data from the {@link InputStream} asynchronously and writes it to the given {@code writableChannel}
+   * @param writableChannel the {@link AsyncWritableChannel} into which data from the {@link InputStream}is written.
+   */
+  private void writeToChannel(final AsyncWritableChannel writableChannel) {
+    executorService.submit(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          final int read = fillBuffer();
+          if (read > 0) {
+            bufferToWrite = ByteBuffer.wrap(buffer);
+            bufferToWrite.limit(read);
+            writableChannel.write(bufferToWrite, new Callback<Long>() {
+              @Override
+              public void onCompletion(Long result, Exception exception) {
+                if (exception != null) {
+                  callbackWrapper.invokeCallback(exception);
+                } else {
+                  callbackWrapper.updateBytesRead(result);
+                  writeToChannel(writableChannel);
+                }
+              }
+            });
+          } else {
+            // the input stream has been emptied.
+            callbackWrapper.invokeCallback(null);
+          }
+        } catch (Exception e) {
+          callbackWrapper.invokeCallback(e);
+        }
+      }
+    });
+  }
+
+  /**
+   * Fills the buffer from the {@link InputStream}.
+   * @return the number of bytes read from the {@link InputStream}.
+   * @throws IOException if there were any problems reading the {@link InputStream}.
+   */
+  private int fillBuffer()
+      throws IOException {
+    int totalRead = 0;
+    int currentRead;
+    do {
+      currentRead = inputStream.read(buffer, totalRead, buffer.length - totalRead);
+      if (currentRead > 0) {
+        totalRead += currentRead;
+      }
+    } while (currentRead < buffer.length && currentRead > 0);
```

Comment:
Also, since `InputStream` never returns 0, `currentRead > 0` is the same as `currentRead !=1`. But I will change it because the latter is probably clearer.